In [2]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
import numpy as np
import math

# Method to find the k-nearest neighbors for a given vector based on cosine similarity.
def find_nearest_neighbors(word_dictionary, vector_1, k, vector_2=None):
  result_dict = {}
  if vector_2 is None:
    for key, value in word_dictionary.items():
      vector_2 = value
      dot_prod = np.dot(vector_1, vector_2)
      norm_vector_1 = math.sqrt(np.dot(vector_1, vector_1))
      norm_vector_2 = math.sqrt(np.dot(vector_2, vector_2))
      cosine_similarility  = dot_prod / (norm_vector_1 * norm_vector_2)
      result_dict[key] = round(cosine_similarility, 3)
    result_dict_sorted = sorted(result_dict.items(), key = lambda x : x[1], reverse=True)
    return result_dict_sorted[1:k + 1]
  else:
    dot_prod = np.dot(vector_1, vector_2)
    norm_vector_1 = math.sqrt(np.dot(vector_1, vector_1))
    norm_vector_2 = math.sqrt(np.dot(vector_2, vector_2))
    cosine_similarility  = dot_prod / (norm_vector_1 * norm_vector_2)
    return cosine_similarility

# Method to generate vector representation.
def generate_vector_representation(word_dictionary, sentence):
  split_arr = sentence.split(' ')
  count = 0
  combined_arr = []
  for val in split_arr:
    val_lower = val.lower()
    if val_lower in word_dictionary:
      vector = word_dictionary[val_lower]
      combined_arr.append(vector)
      count+= 1
    else:
      print("Ignoring the word since it is not found in the dictionary : ", val_lower)
  result_arr = [sum(x)/count for x in zip(*combined_arr)]
  return result_arr

In [0]:
if __name__ == '__main__':
    # read the file.
    input_file = "/content/drive/My Drive/Colab Notebooks/glove.6B.300d.txt"
    word_dictionary = {}
    with open(input_file, 'r', encoding='utf-8') as file:
        for line in file:
            arr = line.split(' ')
            word_dictionary[arr[0]] = np.asfarray(arr[1:], float)

In [0]:
# Task 1.
# Find nearest neighbors to the first name.
task1_result = find_nearest_neighbors(word_dictionary, word_dictionary['kunal'], 5)

In [0]:
# Task 2.
# Find nearest neighbors to a sentence. First generate the vector for the sentence.
sentence1 = "People in US like football"
result_vector = generate_vector_representation(word_dictionary, sentence1)
task2_result = find_nearest_neighbors(word_dictionary, result_vector, 5)

In [0]:
# Task 3.
# Compare 2 sentences.
# similar meaning
sentence2 = "Football is a popular sport in US"
result_vector_1 = generate_vector_representation(word_dictionary, sentence2)
task3_result_1 = find_nearest_neighbors(word_dictionary, result_vector_1, 5, result_vector)

# dissimilar meaning
sentence3 = "Computer Science coursework at Rutgers is excellent"
result_vector_2 = generate_vector_representation(word_dictionary, sentence3)
task3_result_2 = find_nearest_neighbors(word_dictionary, result_vector_2, 5, result_vector)

**Output Task 1:**

In [8]:
print("First Name :- kunal")
print("Nearest Neighbors :- ", task1_result)

First Name :- kunal
Nearest Neighbors :-  [('siddharth', 0.483), ('neha', 0.477), ('kohli', 0.463), ('kapur', 0.462), ('abhay', 0.461)]


The 5 nearest neighbors for my First Name are shown above.

**Output Task 2:**



In [9]:
print("Sentence - ", sentence1)
print("Nearest Neighbors to the sentence :- ", task2_result)

Sentence -  People in US like football
Nearest Neighbors to the sentence :-  [('in', 0.731), ('well', 0.728), ('all', 0.721), ('just', 0.716), ('come', 0.71)]


The 5 nearest neighbors to 'Sentence1' i.e "People in US like football" are shown above.

**Output Task 3:**

In [10]:
# Output for the similar meaning sentence.
print("Sentence 2 :- ", sentence2)
print("Cosine similarity score with Sentence 1 :- ", task3_result_1)

# Output for the dissimilar meaning sentence.
print("Sentence 3 :- ", sentence3)
print("Cosine similarity score with Sentence 1 :- ", task3_result_2)


Sentence 2 :-  Football is a popular sport in US
Cosine similarity score with Sentence 1 :-  0.8705283433798667
Sentence 3 :-  Computer Science coursework at Rutgers is excellent
Cosine similarity score with Sentence 1 :-  0.4948363285260474


The cosine similarity values obtained above make sense as the 'Sentence1' and 'Sentence2' both talk about "Football in United States". So, the angle between the vectors representing these 2 sentences would be very small and the cosine similarity score should be high which is observed in the above results.(~0.87)

Sentence1 talks about Football in United States whereas Sentence3 talks about the Computer Science coursework at Rutgers. The lower cosine similarity score in this case (~0.49) reflects that these 2 sentences are not similar to each other.